pytorch_lightning/trainer/trainer.py

```diff
@@ -1174,6 +1177,18 @@ def call_hook(self, hook_name: str, *args, **kwargs) -> Any:
 
         return output
 
+    def _parse_devices(
```

Comment:
does this function really have to live in the trainer? Can't it completely go to utils?

Reply:
The motivation behind it is we had a use case where they would like to train the models remotely, and don't have access to the machines locally. This would cause an exception during the Trainer instantiation itself, with no available local devices.

This PR will help the users to override the device parsing functionality and build on top of Trainer with flexibility.

Reply:
But then you are intending to make it part of the public interface, so at least the underscore has to be removed :)

Reply:
But don't want users to get an idea or are advised to access the `parse_devices` functionality through Trainer.

Reply:
actually, I kind of like the idea of having it publicly available. Maybe something like

```python
@staticmethod
def available_devices():
    ...
    return {'gpus': [0, 1, 2], 'cpu': 16} 
```

Because currently I most of the time have checks beforehand like 
```python
if torch.cuda.is_available(): 
    n_gpus = torch.cuda.device_count()
elif SOME_XLA_XONDITION:
   n_tpus = ...

```

Reply:
@justusschock Are you suggesting to add `available_devices` method? Tbh, I would like it better than `parse_devices` raises a Warning when the devices are not available.

Reply:
One thing doesn't mean you can't do the other.

I think we definitely should add a warning if the device isn't available, but having such a method could also allow to write more agnostic code without having to do the same checks manually :)

Reply:
@justusschock For sure! Something we could consider. This is a refactor PR for now focused on decoupling some logic, we could introduce additional logic in a follow-up PR.